### PR TITLE
[Merged by Bors] - feat(Order/WithBot): add `forall_iff_eq_bot ` and `forall_iff_eq_top `

### PR DESCRIPTION
**Pull request by pitmonticone**
- [x] Add `forall_iff_eq_bot `
- [x] Add `forall_iff_eq_top `

Co-authored-by: @D-Thomine 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by pitmonticone** [comment]
Thank you for you review comments @fpvandoorn. 

`Option.ne_none_iff_exists` does not seem to be in Mathlib. Where should I put the new lemma you're asking for?

File: Mathlib/Order/WithBot.lean [hunk 1]
```diff
@@ -1067,6 +1067,16 @@ theorem ofDual_map (f : αᵒᵈ → βᵒᵈ) (a : WithBot αᵒᵈ) :
   rfl
 #align with_bot.of_dual_map WithBot.ofDual_map
 
+theorem forall_iff_eq_bot [Preorder α] {x : WithBot α} :
```
[the remark this answers is not in the document]
Comment (pitmonticone) [reply Mathlib/Order/WithBot.lean]: I'm not at all sure how to proceed here.

**Comment by fpvandoorn** [comment]
I would add the `Iff.not` versions of `WithBot.ne_bot_iff_exists` and `Option.ne_none_iff_exists` just below the existing versions (for `Option` it can live in `Data/Option/Basic`.

**Comment by pitmonticone** [comment]
I should have addressed all the comments and suggestions.

Thank you @fpvandoorn for your review!

File: Mathlib/Order/WithBot.lean [hunk 2]
```diff
@@ -1067,6 +1069,14 @@ theorem ofDual_map (f : αᵒᵈ → βᵒᵈ) (a : WithBot αᵒᵈ) :
   rfl
 #align with_bot.of_dual_map WithBot.ofDual_map
 
+lemma forall_lt_iff_eq_bot [Preorder α] {x : WithBot α} :
+    (∀ y : α, x < y) ↔ x = ⊥ := by
+  refine ⟨fun h ↦ ?_, fun h ↦ h ▸ fun y ↦ WithBot.bot_lt_coe y⟩
+  by_contra h'
```
Comment (fpvandoorn) [inline Mathlib/Order/WithBot.lean:1075]: This is now very short with `eq_bot_iff_forall`, right?
Something like (untested) `exact eq_top_iff_forall.mpr (fun x \mapsto (h x).ne')`

Comment (fpvandoorn) [reply Mathlib/Order/WithBot.lean:1075]: Maybe it would also be nice to unformize whether `x = ⊥` is LHS or RHS, but I don't have a strong opinion there.

Comment (pitmonticone) [reply Mathlib/Order/WithBot.lean:1075]: Done. Thanks.

**Comment by pitmonticone** [comment]
I should have addressed all the comments and suggestions.

Thank you @fpvandoorn for your review!

**Comment by pitmonticone** [comment]
I should have addressed all the comments and suggestions.

Thank you @fpvandoorn for your review!

File: Mathlib/Data/Option/Basic.lean [hunk 3]
```diff
@@ -477,4 +477,8 @@ lemma isNone_eq_false_iff (a : Option α) : Option.isNone a = false ↔ Option.i
 lemma eq_none_or_eq_some (a : Option α) : a = none ∨ ∃ x, a = some x :=
   Option.exists.mp exists_eq'
 
+lemma eq_none_iff_forall {o : Option α} : (∀ (x : α), some x ≠ o) ↔ o = none := by
```
Comment (fpvandoorn) [inline Mathlib/Data/Option/Basic.lean:480]:
```suggestion
lemma forall_ne_iff_eq_none {o : Option α} : (∀ (x : α), some x ≠ o) ↔ o = none := by
```
(or perhaps `forall_some_ne_iff_eq_none`). Similar in the other new `ne` lemmas.

Thanks!

bors d+

Comment (pitmonticone) [reply Mathlib/Data/Option/Basic.lean:480]: Ops, you're right. Sorry. I'll do it later.

**Pull request closed without merging**